leader/leader.go

```diff
@@ -105,279 +94,275 @@ func (l *Client) AddWatchCallback(key string, retry time.Duration, fn CallbackFn
+				logger.WithError(err).Error("Unexpected watch error.")
+				if steps > 10 {
+					watcher = nil
+					b.Reset()
+					steps = 0
 				}
 			}
-
+			return true
+		}
+	}
+	for tick(err) {
+		if watcher == nil {
+			watcher, err = l.getWatchAtLatestIndex(ctx, api, key, valuesC)
 			if err != nil {
-				select {
-				case <-ticker.C:
-					steps += 1
-				}
-
-				if err == context.Canceled {
-					return
-				} else if cerr, ok := err.(*client.ClusterError); ok {
-					if len(cerr.Errors) != 0 && cerr.Errors[0] == context.Canceled {
-						return
-					}
-					log.Debugf("unexpected cluster error: %v (%v)", err, cerr.Detail())
-					continue
-				} else if IsWatchExpired(err) {
-					log.Debug("watch expired, resetting watch index")
-					watcher, resp, err = l.getWatchAtLatestIndex(ctx, api, key, retry)
-					if err != nil {
-						continue
-					}
-				} else {
-					log.Warningf("unexpected watch error: %v", err)
-					// try recreating the watch if we get repeated unknown errors
-					if steps > 10 {
-						watcher, resp, err = l.getWatchAtLatestIndex(ctx, api, key, retry)
-						if err != nil {
-							continue
-						}
-						backoff.Reset()
-						steps = 0
-					}
-
-					continue
-				}
-			}
-			// if nothing has changed and we previously sent this subscriber this value,
-			// do not bother subscriber with extra notifications
-			if resp.PrevNode != nil && resp.PrevNode.Value == resp.Node.Value && sentAnything {
 				continue
 			}
-			select {
-			case valuesC <- resp.Node.Value:
-				sentAnything = true
-			case <-l.closeC:
-				return
-			}
+			// Successful return means the current value has been sent to receiver
 		}
-	}()
+		resp, err = watcher.Next(ctx)
+		if err != nil {
+			continue
+		}
+		if resp.Node.Value == "" {
+			continue
+		}
+		if resp.PrevNode != nil && resp.PrevNode.Value == resp.Node.Value {
+			continue
+		}
+		select {
+		case valuesC <- resp.Node.Value:
+		case <-l.closeC:
+			logger.Info("Watcher is closing.")
+			return
+		}
+	}
 }
 
-// AddVoter starts a goroutine that attempts to set the specified key to
+// voterLoop is a process that attempts to set the specified key to
 // to the given value with the time-to-live value specified with term.
 // The time-to-live value cannot be less than a second.
 // After successfully setting the key, it attempts to renew the lease for the specified
-// term indefinitely
-func (l *Client) AddVoter(context context.Context, key, value string, term time.Duration) error {
-	if value == "" {
-		return trace.BadParameter("voter value for key cannot be empty")
-	}
-	if term < time.Second {
-		return trace.BadParameter("term cannot be < 1s")
-	}
+// term indefinitely.
+func (l *Client) voterLoop(key, value string, term time.Duration, enabled bool) {
+	logger := log.WithFields(logrus.Fields{
+		"key":   key,
+		"value": value,
+		"term":  term,
+	})
+	var ticker *time.Ticker
+	var tickerC <-chan time.Time
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
Same question as above - would it make sense to pass the context from the caller?

Reply:
It's sort of the intent since the context is local to this method - otherwise, it would be confusing to see context **and** selecting on `l.closeC`.